# <center>HW 1: Data Cleaning and Wrangling</center>
## <center>CS/CE 457/464: Data Science</center>
## <center>Samiya Ali Zaidi</center>
### <center>sa07171@st.habib.edu.pk</center>

# Part 1: Data Formatting

In this section, I will format the data to remove redundacy, and for keeping one column for one individual question, because what we have right now can not be used at all.

In [119]:
import pandas as pd

In [121]:
data = pd.read_csv('580SurveyCleanup2.csv')
data.head()

Unnamed: 0       Unnamed: 1 Unnamed: 2          Unnamed: 3   Unnamed: 4  \
0        No.             Time     Status  Duration (Seconds)  Access Code   
1          1  10/3/2023 17:49  Completed               11646          NaN   
2          2  10/3/2023 17:49  Completed                 113          NaN   
3          3  10/3/2023 17:49  Completed                 572          NaN   
4          4  10/3/2023 17:50  Completed                 365          NaN   

      Unnamed: 5      Unnamed: 6     Unnamed: 7 Unnamed: 8  \
0  Email Address      IP Address        Country     Region   
1            NaN  100.36.167.226  United States   Virginia   
2            NaN   98.169.26.134  United States   Virginia   
3            NaN    98.169.52.23  United States   Virginia   
4            NaN  69.255.182.253  United States   Virginia   

               Unnamed: 9  ...             Q12.1         Q12.2  \
0  City / Town / District  ...  Some familiarity  Average user   
1                 Fairfax  ...                 1           NaN   
2                 Fairfax  ...               NaN           NaN   
3          Greenway Downs  ...               NaN             1   
4             Gainesville  ...                 1           NaN   

                       Q12.3          Q12.4  Q13 Q13.1  Q14 Q14.1     Q14.2  \
0  Frequent use for projects  Fluent/expert  Yes    No  Yes    No  Not sure   
1                        NaN            NaN    1   NaN    1   NaN       NaN   
2                        NaN            NaN    1   NaN  NaN     1       NaN   
3                        NaN            NaN  NaN     1  NaN   NaN         1   
4                        NaN            NaN  NaN     1    1   NaN       NaN   

                                               Q13.2  
0  What is your goal for learning about data anal...  
1                                                NaN  
2                                                NaN  
3  As part of my work, I have access to a lot of ...  
4                                                NaN  

[5 rows x 51 columns]

## Replacing the initial headers

In [124]:
df_gen = data.iloc[:,1:12]
df_gen.columns = df_gen.iloc[0]
df_gen = df_gen[1:]
df_gen.head()

0             Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

0      IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

0 Longitude  
1   38.8209  
2   38.8605  
3   38.9645  
4   38.8175  
5    37.751

## Cleaning the headers for the actual questions asked in the survey.

### For Q1

In [128]:
df_survey = data.iloc[:,[12,13]]

In [130]:
df_survey.columns = df_survey.iloc[0].astype(str)
df_survey = df_survey[1:]

In [132]:
df_survey = df_survey.loc[:, '2.0':'4.0'].replace(1, pd.Series(df_survey.columns, df_survey.columns)).fillna('')
df_survey.head()

0  2.0  4.0
1       4.0
2  2.0     
3  2.0     
4  2.0     
5  2.0

In [134]:
df_survey = pd.DataFrame(df_survey.iloc[:,0] + df_survey.iloc[:,1])
df_survey.columns = ['Section']
df_survey.head()

Section
1     4.0
2     2.0
3     2.0
4     2.0
5     2.0

In [136]:
# add to the df_gen 
df_gen = pd.concat([df_gen, df_survey],axis=1)
df_gen.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

  Longitude Section  
1   38.8209     4.0  
2   38.8605     2.0  
3   38.9645     2.0  
4   38.8175     2.0  
5    37.751     2.0

### For Q2

In [9]:
# getting headers for the gender question
df_q2 = data.iloc[:,[14,15, 16, 17]]
df_q2.head()

Q2    Q2.1       Q2.2                  Q2.3
0  Male  Female  NonBinary  Prefer not to answer
1   NaN       1        NaN                   NaN
2   NaN       1        NaN                   NaN
3     1     NaN        NaN                   NaN
4   NaN     NaN        NaN                     1

In [10]:
# replacing with first column
df_q2.columns = df_q2.iloc[0].astype(str)
df_q2 = df_q2[1:]  

In [11]:
df_q2.head()

0 Male Female NonBinary Prefer not to answer
1  NaN      1       NaN                  NaN
2  NaN      1       NaN                  NaN
3    1    NaN       NaN                  NaN
4  NaN    NaN       NaN                    1
5    1    NaN       NaN                  NaN

In [12]:
df_q2 = df_q2.iloc[:, 0:4].replace('1', pd.Series(df_q2.columns, df_q2.columns)).fillna('')

In [13]:
df_q2.head()

0  Male  Female NonBinary  Prefer not to answer
1        Female                                
2        Female                                
3  Male                                        
4                          Prefer not to answer
5  Male

In [14]:
df_q2 = pd.DataFrame(df_q2.iloc[:,0] + df_q2.iloc[:,1] + df_q2.iloc[:,2] + df_q2.iloc[:,3])

In [15]:
df_q2.columns = ['Gender']

In [16]:
df_q2.head()

Gender
1                Female
2                Female
3                  Male
4  Prefer not to answer
5                  Male

In [17]:
df_q2.iloc[:, 0] = df_q2.iloc[:, 0].replace('Female', 'F')
df_q2.iloc[:, 0] = df_q2.iloc[:, 0].replace('Male', 'M')
df_q2.iloc[:, 0] = df_q2.iloc[:, 0].replace('NonBinary', 'NB')
df_q2.iloc[:, 0] = df_q2.iloc[:, 0].replace('Prefer not to answer', 'No')

df_q2.head()

Gender
1      F
2      F
3      M
4     No
5      M

In [18]:
# add to the main df_gen
df_gen = pd.concat([df_gen, df_q2], axis=1)
df_gen.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

  Longitude Section Gender  
1   38.8209     4.0      F  
2   38.8605     2.0      F  
3   38.9645     2.0      M  
4   38.8175     2.0     No  
5    37.751     2.0      M

### For Q3

In [19]:
df_q3 = data.iloc[:, [18]]

In [20]:
# rename 
df_q3.columns = df_q3.iloc[0].astype(str)
df_q3 = df_q3[1:]  

In [21]:
df_q3.columns = ['Age']

In [22]:
df_q3.head()

Age
1  23
2  24
3  40
4  26
5  30

In [23]:
df_gen = pd.concat([df_gen, df_q3], axis=1)
df_gen.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

  Longitude Section Gender Age  
1   38.8209     4.0      F  23  
2   38.8605     2.0      F  24  
3   38.9645     2.0      M  40  
4   38.8175     2.0     No  26  
5    37.751     2.0      M  30

### For Q4 - Q7

In [24]:
df_q4_7 = data.iloc[:,[19,20,21,22]]

In [25]:
df_q4_7.columns = df_q4_7.iloc[0].astype(str)
df_q4_7 = df_q4_7[1:]  

In [26]:
df_q4_7.columns = ['Height', 'Citizenship', 'Undergrad_Degree', 'GradYear']

In [27]:
df_gen = pd.concat([df_gen, df_q4_7], axis=1)

In [28]:
df_gen.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

  Longitude Section Gender Age Height  Citizenship       Undergrad_Degree  \
1   38.8209     4.0      F  23     63       Taiwan              economics   
2   38.8605     2.0      F  24     59        India               Finance    
3   38.9645     2.0      M  40     70        India  Computer Applications   
4   38.8175     2.0     No  26     66  South Korea    Business Management   
5    37.751     2.0      M  30     70           US                    Yes   

     GradYear  
1  12/31/2021  
2    5/1/2021  
3    6/1/2022  
4  12/31/2021  
5    1/1/2021

### For Q8

In [29]:
df_q8 = data.iloc[:, [23, 24, 25]]

In [30]:
df_q8.columns = df_q8.iloc[0].astype(str)
df_q8 = df_q8[1:]  

In [31]:
df_q8['Other'].count()

1

This means I shouldn't drop the other column.

In [32]:
df_q8.columns = ['Windows', 'MacBook', 'Other']

In [33]:
for col in df_q8.columns:
    df_q8[col] = df_q8[col].replace({'1': col, '2': col})
df_q8 = df_q8.replace('0', '').fillna('')

In [34]:
df_q8 = pd.DataFrame(df_q8.iloc[:,0] + df_q8.iloc[:,1] + df_q8.iloc[:,2])

In [35]:
df_q8.columns = ['Laptop']

In [36]:
df_q8.head()

Laptop
1  MacBook
2  MacBook
3  MacBook
4  Windows
5  MacBook

In [37]:
df_gen = pd.concat([df_gen, df_q8], axis=1)

In [38]:
df_gen.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

  Longitude Section Gender Age Height  Citizenship       Undergrad_Degree  \
1   38.8209     4.0      F  23     63       Taiwan              economics   
2   38.8605     2.0      F  24     59        India               Finance    
3   38.9645     2.0      M  40     70        India  Computer Applications   
4   38.8175     2.0     No  26     66  South Korea    Business Management   
5    37.751     2.0      M  30     70           US                    Yes   

     GradYear   Laptop  
1  12/31/2021  MacBook  
2    5/1/2021  MacBook  
3    6/1/2022  MacBook  
4  12/31/2021  Windows  
5    1/1/2021  MacBook

### For Q9

In [39]:
df_q9 = data.iloc[:, [27, 28, 29]]

In [40]:
df_q9.columns = df_q9.iloc[0].astype(str)
df_q9 = df_q9[1:]  

In [41]:
df_q9.columns = ['Full Time', 'Part Time', 'Not Working']

In [42]:
df_q9 = df_q9.loc[:, 'Full Time':'Not Working'].replace('1', pd.Series(df_q9.columns, df_q9.columns)).fillna('')

In [43]:
df_q9 = pd.DataFrame(df_q9.iloc[:,0] + df_q9.iloc[:,1] + df_q9.iloc[:,2])

In [44]:
df_q9.columns = ['Employement']

In [45]:
df_q9.head()

Employement
1  Not Working
2    Part Time
3    Full Time
4  Not Working
5    Full Time

In [46]:
df_gen = pd.concat([df_gen, df_q9], axis=1)

In [47]:
df_gen.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

  Longitude Section Gender Age Height  Citizenship       Undergrad_Degree  \
1   38.8209     4.0      F  23     63       Taiwan              economics   
2   38.8605     2.0      F  24     59        India               Finance    
3   38.9645     2.0      M  40     70        India  Computer Applications   
4   38.8175     2.0     No  26     66  South Korea    Business Management   
5    37.751     2.0      M  30     70           US                    Yes   

     GradYear   Laptop  Employement  
1  12/31/2021  MacBook  Not Working  
2    5/1/2021  MacBook    Part Time  
3    6/1/2022  MacBook    Full Time  
4  12/31/2021  Windows  Not Working  
5    1/1/2021  MacBook    Full Time

### For Q10

In [48]:
df_q10 = data.iloc[:,[30, 31, 32, 33, 34]]

In [49]:
df_q10.columns = df_q10.iloc[0].astype(str)
df_q10 = df_q10[1:]  

In [50]:
df_q10 = df_q10.loc[:, 'Little/none':'Fluent/expert'].replace('1', pd.Series(df_q10.columns, df_q10.columns)).fillna('')

In [51]:
df_q10 = pd.DataFrame(df_q10.iloc[:,0] + df_q10.iloc[:,1] + df_q10.iloc[:,2] + df_q10.iloc[:,3] + df_q10.iloc[:,4])

Instead of keeping the text for the skill level, I wil convert it into 1 - 5

In [52]:
df_q10.iloc[:, 0] = df_q10.iloc[:, 0].replace('Little/none', 1)
df_q10.iloc[:, 0] = df_q10.iloc[:, 0].replace('Some familiarity', 2)
df_q10.iloc[:, 0] = df_q10.iloc[:, 0].replace('Average user', 3)
df_q10.iloc[:, 0] = df_q10.iloc[:, 0].replace('Frequent use for projects', 4)
df_q10.iloc[:, 0] = df_q10.iloc[:, 0].replace('Fluent/expert', 5)

In [53]:
df_q10.columns = ['Python Skills']

In [54]:
df_q10.head()

Python Skills
1             2
2             3
3             2
4             3
5             2

In [55]:
df_gen = pd.concat([df_gen, df_q10], axis=1)

In [56]:
df_gen.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

   ... Section Gender Age Height  Citizenship       Undergrad_Degree  \
1  ...     4.0      F  23     63       Taiwan              economics   
2  ...     2.0      F  24     59        India               Finance    
3  ...     2.0      M  40     70        India  Computer Applications   
4  ...     2.0     No  26     66  South Korea    Business Management   
5  ...     2.0      M  30     70           US                    Yes   

     GradYear   Laptop  Employement Python Skills  
1  12/31/2021  MacBook  Not Working             2  
2    5/1/2021  MacBook    Part Time             3  
3    6/1/2022  MacBook    Full Time             2  
4  12/31/2021  Windows  Not Working             3  
5    1/1/2021  MacBook    Full Time             2  

[5 rows x 21 columns]

### For Q11

In [57]:
df_q11 = data.iloc[:,[35, 36, 37, 38, 39]]

In [58]:
df_q11.head()

Q11             Q11.1         Q11.2                      Q11.3  \
0  Little/none  Some familiarity  Average user  Frequent use for projects   
1            1               NaN           NaN                        NaN   
2          NaN               NaN             1                        NaN   
3            1               NaN           NaN                        NaN   
4          NaN               NaN           NaN                          1   

           Q11.4  
0  Fluent/expert  
1            NaN  
2            NaN  
3            NaN  
4            NaN

In [59]:
df_q11.columns = df_q11.iloc[0].astype(str)
df_q11 = df_q11[1:]  

In [60]:
df_q11 = df_q11.loc[:, 'Little/none':'Fluent/expert'].replace('1', pd.Series(df_q11.columns, df_q11.columns)).fillna('')

In [61]:
df_q11.head()

0  Little/none  Some familiarity  Average user  Frequent use for projects  \
1  Little/none                                                              
2                                 Average user                              
3  Little/none                                                              
4                                               Frequent use for projects   
5               Some familiarity                                            

0 Fluent/expert  
1                
2                
3                
4                
5

In [62]:
df_q11 = pd.DataFrame(df_q11.iloc[:,0] + df_q11.iloc[:,1] + df_q11.iloc[:,2] + df_q11.iloc[:,3] + df_q11.iloc[:,4])

In [63]:
df_q11.iloc[:, 0] = df_q11.iloc[:, 0].replace('Little/none', 1)
df_q11.iloc[:, 0] = df_q11.iloc[:, 0].replace('Some familiarity', 2)
df_q11.iloc[:, 0] = df_q11.iloc[:, 0].replace('Average user', 3)
df_q11.iloc[:, 0] = df_q11.iloc[:, 0].replace('Frequent use for projects', 4)
df_q11.iloc[:, 0] = df_q11.iloc[:, 0].replace('Fluent/expert', 5)

In [64]:
df_q11.columns = ['R Skills']

In [65]:
df_q11.head()

R Skills
1        1
2        3
3        1
4        4
5        2

In [66]:
df_gen = pd.concat([df_gen, df_q11], axis=1)

In [67]:
df_gen.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

   ... Gender Age Height  Citizenship       Undergrad_Degree    GradYear  \
1  ...      F  23     63       Taiwan              economics  12/31/2021   
2  ...      F  24     59        India               Finance     5/1/2021   
3  ...      M  40     70        India  Computer Applications    6/1/2022   
4  ...     No  26     66  South Korea    Business Management  12/31/2021   
5  ...      M  30     70           US                    Yes    1/1/2021   

    Laptop  Employement Python Skills R Skills  
1  MacBook  Not Working             2        1  
2  MacBook    Part Time             3        3  
3  MacBook    Full Time             2        1  
4  Windows  Not Working             3        4  
5  MacBook    Full Time             2        2  

[5 rows x 22 columns]

### For Q12

In [68]:
df_q12 = data.iloc[:, [40, 41, 42, 43, 44]]

In [69]:
df_q12.columns = df_q12.iloc[0].astype(str)
df_q12 = df_q12[1:]

In [70]:
df_q12 = df_q12.loc[:, 'Little/none':'Fluent/expert'].replace('1', pd.Series(df_q12.columns, df_q12.columns)).fillna('')

In [71]:
df_q12 = pd.DataFrame(df_q12.iloc[:,0] + df_q12.iloc[:,1] + df_q12.iloc[:,2] + df_q12.iloc[:,3] + df_q12.iloc[:,4])

In [72]:
df_q12.columns = ['SQL Skills']

In [73]:
df_q12.iloc[:, 0] = df_q12.iloc[:, 0].replace('Little/none', 1)
df_q12.iloc[:, 0] = df_q12.iloc[:, 0].replace('Some familiarity', 2)
df_q12.iloc[:, 0] = df_q12.iloc[:, 0].replace('Average user', 3)
df_q12.iloc[:, 0] = df_q12.iloc[:, 0].replace('Frequent use for projects', 4)
df_q12.iloc[:, 0] = df_q12.iloc[:, 0].replace('Fluent/expert', 5)

In [74]:
df_q12.head()

SQL Skills
1          2
2          1
3          3
4          2
5          3

In [75]:
df_gen = pd.concat([df_gen, df_q12], axis=1)

In [76]:
df_gen.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

   ... Age Height  Citizenship       Undergrad_Degree    GradYear   Laptop  \
1  ...  23     63       Taiwan              economics  12/31/2021  MacBook   
2  ...  24     59        India               Finance     5/1/2021  MacBook   
3  ...  40     70        India  Computer Applications    6/1/2022  MacBook   
4  ...  26     66  South Korea    Business Management  12/31/2021  Windows   
5  ...  30     70           US                    Yes    1/1/2021  MacBook   

   Employement Python Skills R Skills SQL Skills  
1  Not Working             2        1          2  
2    Part Time             3        3          1  
3    Full Time             2        1          3  
4  Not Working             3        4          2  
5    Full Time             2        2          3  

[5 rows x 23 columns]

### For Q13

In [77]:
df_q13 = data.iloc[:, [45, 46]]

In [78]:
df_q13.columns = df_q13.iloc[0].astype(str)
df_q13 = df_q13[1:]

In [79]:
df_q13 = df_q13.loc[:, 'Yes':'No'].replace('1', pd.Series(df_q13.columns, df_q13.columns)).fillna('')

In [80]:
df_q13 = pd.DataFrame(df_q13.iloc[:,0] + df_q13.iloc[:,1])

Instead of using strings, it would be better to use 1s and 0s to make further analysis easy.

In [81]:
df_q13.iloc[:, 0] = df_q13.iloc[:, 0].replace('Yes', 1)
df_q13.iloc[:, 0] = df_q13.iloc[:, 0].replace('No', 0)

In [82]:
df_q13.columns = ['GitHub']

In [83]:
df_q13.head()

GitHub
1      1
2      1
3      0
4      0
5      0

In [84]:
df_gen = pd.concat([df_gen, df_q13], axis=1)

In [85]:
df_gen.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

   ... Height  Citizenship       Undergrad_Degree    GradYear   Laptop  \
1  ...     63       Taiwan              economics  12/31/2021  MacBook   
2  ...     59        India               Finance     5/1/2021  MacBook   
3  ...     70        India  Computer Applications    6/1/2022  MacBook   
4  ...     66  South Korea    Business Management  12/31/2021  Windows   
5  ...     70           US                    Yes    1/1/2021  MacBook   

   Employement Python Skills R Skills SQL Skills GitHub  
1  Not Working             2        1          2      1  
2    Part Time             3        3          1      1  
3    Full Time             2        1          3      0  
4  Not Working             3        4          2      0  
5    Full Time             2        2          3      0  

[5 rows x 24 columns]

### For Q14

In [86]:
df_q14 = data.iloc[:, [47, 48, 49]]

In [87]:
df_q14.columns = df_q14.iloc[0].astype(str)
df_q14 = df_q14[1:]

In [88]:
df_q14 = df_q14.loc[:, 'Yes':'Not sure'].replace('1', pd.Series(df_q14.columns, df_q14.columns)).fillna('')

In [89]:
df_q14 = pd.DataFrame(df_q14.iloc[:, 0] + df_q14.iloc[:, 1] + df_q14.iloc[:, 2])

In [90]:
df_q14.columns = ['Career in DS']

In [91]:
df_q14.head()

Career in DS
1          Yes
2           No
3     Not sure
4          Yes
5          Yes

In [92]:
df_gen = pd.concat([df_gen, df_q14], axis=1)

In [93]:
df_gen.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

   ...  Citizenship       Undergrad_Degree    GradYear   Laptop  Employement  \
1  ...       Taiwan              economics  12/31/2021  MacBook  Not Working   
2  ...        India               Finance     5/1/2021  MacBook    Part Time   
3  ...        India  Computer Applications    6/1/2022  MacBook    Full Time   
4  ...  South Korea    Business Management  12/31/2021  Windows  Not Working   
5  ...           US                    Yes    1/1/2021  MacBook    Full Time   

  Python Skills R Skills SQL Skills GitHub Career in DS  
1             2        1          2      1          Yes  
2             3        3          1      1           No  
3             2        1          3      0     Not sure  
4             3        4          2      0          Yes  
5             2        2          3      0          Yes  

[5 rows x 25 columns]

### For Q15

In [94]:
df_q15 = data.iloc[:, [50]]

In [95]:
df_q15.columns = ['Goals']
df_q15 = df_q15[1:]

In [96]:
df_q15 = df_q15.loc[:, 'Goals':].replace(1, pd.Series(df_q15.columns, df_q15.columns)).fillna('')

In [97]:
df_q15.head()

Goals
1                                                   
2                                                   
3  As part of my work, I have access to a lot of ...
4                                                   
5                                              Learn

In [98]:
df_gen = pd.concat([df_gen, df_q15], axis=1)

In [99]:
df_gen.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/2023 17:49  Completed              11646         NaN           NaN   
2  10/3/2023 17:49  Completed                113         NaN           NaN   
3  10/3/2023 17:49  Completed                572         NaN           NaN   
4  10/3/2023 17:50  Completed                365         NaN           NaN   
5  10/3/2023 17:51  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

   ...       Undergrad_Degree    GradYear   Laptop  Employement Python Skills  \
1  ...              economics  12/31/2021  MacBook  Not Working             2   
2  ...               Finance     5/1/2021  MacBook    Part Time             3   
3  ...  Computer Applications    6/1/2022  MacBook    Full Time             2   
4  ...    Business Management  12/31/2021  Windows  Not Working             3   
5  ...                    Yes    1/1/2021  MacBook    Full Time             2   

  R Skills SQL Skills GitHub Career in DS  \
1        1          2      1          Yes   
2        3          1      1           No   
3        1          3      0     Not sure   
4        4          2      0          Yes   
5        2          3      0          Yes   

                                               Goals  
1                                                     
2                                                     
3  As part of my work, I have access to a lot of ...  
4                                                     
5                                              Learn  

[5 rows x 26 columns]

## Final Formatted Data

The dataset has been reduced from 51 columns originally to 26 formatted columns.

In [100]:
df_gen.head(10)

Time      Status Duration (Seconds) Access Code Email Address  \
1   10/3/2023 17:49   Completed              11646         NaN           NaN   
2   10/3/2023 17:49   Completed                113         NaN           NaN   
3   10/3/2023 17:49   Completed                572         NaN           NaN   
4   10/3/2023 17:50   Completed                365         NaN           NaN   
5   10/3/2023 17:51   Completed                 92         NaN           NaN   
6   10/3/2023 17:59  Incomplete                221         NaN           NaN   
7   10/3/2023 18:23   Completed             115983         NaN           NaN   
8   10/3/2023 19:10  Incomplete                 40         NaN           NaN   
9   10/3/2023 19:25   Completed                137         NaN           NaN   
10  10/3/2023 19:56  Incomplete                179         NaN           NaN   

        IP Address        Country    Region City / Town / District  Latitude  \
1   100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2    98.169.26.134  United States  Virginia                Fairfax   -77.263   
3     98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4   69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5   138.88.108.230  United States       NaN                    NaN   -97.822   
6    108.56.134.80  United States  Virginia                 Reston  -77.3489   
7   108.48.175.220  United States  Virginia              Arlington  -77.0987   
8    173.67.199.85  United States  Virginia                Ashburn   -77.539   
9     74.96.114.12  United States  Virginia             Woodbridge  -77.3197   
10   107.77.241.32  United States     Texas                    NaN  -96.8217   

    ...         Undergrad_Degree    GradYear   Laptop  Employement  \
1   ...                economics  12/31/2021  MacBook  Not Working   
2   ...                 Finance     5/1/2021  MacBook    Part Time   
3   ...    Computer Applications    6/1/2022  MacBook    Full Time   
4   ...      Business Management  12/31/2021  Windows  Not Working   
5   ...                      Yes    1/1/2021  MacBook    Full Time   
6   ...                      NaN         NaN                         
7   ...                  Finance   7/15/2022  Windows    Part Time   
8   ...                      NaN         NaN                         
9   ...                 Economic  12/15/2021  Windows    Full Time   
10  ...  Economics with Finance          NaN                         

   Python Skills R Skills SQL Skills GitHub Career in DS  \
1              2        1          2      1          Yes   
2              3        3          1      1           No   
3              2        1          3      0     Not sure   
4              3        4          2      0          Yes   
5              2        2          3      0          Yes   
6                                         1           No   
7              3        3          3      0           No   
8                                         1           No   
9              2        4          3      1           No   
10                                        0           No   

                                                Goals  
1                                                      
2                                                      
3   As part of my work, I have access to a lot of ...  
4                                                      
5                                               Learn  
6                                                      
7   To deduce meaningful interpretation for Financ...  
8                                                      
9                                                      
10                                                     

[10 rows x 26 columns]

In [101]:
# saving to a CSV file
df_gen.to_csv('formatted_hw1_data.csv', index=False)

# Part 2: Data Cleaning

Make a copy for the originally formatted data.

In [102]:
df_clean = pd.read_csv('formatted_hw1_data.csv')

In [103]:
df_clean.head()

Time     Status  Duration (Seconds)  Access Code  Email Address  \
0  10/3/2023 17:49  Completed               11646          NaN            NaN   
1  10/3/2023 17:49  Completed                 113          NaN            NaN   
2  10/3/2023 17:49  Completed                 572          NaN            NaN   
3  10/3/2023 17:50  Completed                 365          NaN            NaN   
4  10/3/2023 17:51  Completed                  92          NaN            NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
0  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
1   98.169.26.134  United States  Virginia                Fairfax  -77.2630   
2    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
3  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
4  138.88.108.230  United States       NaN                    NaN  -97.8220   

   ...       Undergrad_Degree    GradYear   Laptop  Employement Python Skills  \
0  ...              economics  12/31/2021  MacBook  Not Working           2.0   
1  ...               Finance     5/1/2021  MacBook    Part Time           3.0   
2  ...  Computer Applications    6/1/2022  MacBook    Full Time           2.0   
3  ...    Business Management  12/31/2021  Windows  Not Working           3.0   
4  ...                    Yes    1/1/2021  MacBook    Full Time           2.0   

  R Skills SQL Skills GitHub Career in DS  \
0      1.0        2.0      1          Yes   
1      3.0        1.0      1           No   
2      1.0        3.0      0     Not sure   
3      4.0        2.0      0          Yes   
4      2.0        3.0      0          Yes   

                                               Goals  
0                                                NaN  
1                                                NaN  
2  As part of my work, I have access to a lot of ...  
3                                                NaN  
4                                              Learn  

[5 rows x 26 columns]

In [104]:
df_clean.columns

Index(['Time', 'Status', 'Duration (Seconds)', 'Access Code', 'Email Address',
       'IP Address', 'Country', 'Region', 'City / Town / District', 'Latitude',
       'Longitude', 'Section', 'Gender', 'Age', 'Height', 'Citizenship',
       'Undergrad_Degree', 'GradYear', 'Laptop', 'Employement',
       'Python Skills', 'R Skills', 'SQL Skills', 'GitHub', 'Career in DS',
       'Goals'],
      dtype='object')

In [105]:
df_clean.isnull().sum()

Time                       0
Status                     0
Duration (Seconds)         0
Access Code               36
Email Address             36
IP Address                 0
Country                    0
Region                     4
City / Town / District     5
Latitude                   0
Longitude                  0
Section                    0
Gender                     0
Age                        0
Height                     1
Citizenship                2
Undergrad_Degree           2
GradYear                   3
Laptop                     4
Employement                3
Python Skills              3
R Skills                   3
SQL Skills                 3
GitHub                     0
Career in DS               0
Goals                      9
dtype: int64

In this dataset, I can observe that two columns, Access Code & Email Address are completely empty, so they should probably be removed from the dataset to remove clutter. Furthermore, there are a few seemingly unnecessary columns as well such as the IP, Time, Duration, Latitude, Longitude, Height. 
While the 'Country' column might not be useless, but it is unnecessary since everyone has the same response, so it should also be dropped from the dataset. 
Since 'Height' is part of the original survey, I won't be dropping it eventhough I think that it's unnecessary.

In [106]:
columns_to_drop = ['Time', 'Duration (Seconds)', 'Access Code', 'Email Address', 'IP Address', 'Country', 'Longitude', 'Latitude']

In [107]:
df_clean = df_clean.drop(columns=columns_to_drop)

In [108]:
df_clean

Status    Region City / Town / District  Section Gender   Age Height  \
0    Completed  Virginia                Fairfax      4.0      F    23     63   
1    Completed  Virginia                Fairfax      2.0      F    24     59   
2    Completed  Virginia         Greenway Downs      2.0      M    40     70   
3    Completed  Virginia            Gainesville      2.0     No    26     66   
4    Completed       NaN                    NaN      2.0      M    30     70   
5   Incomplete  Virginia                 Reston      2.0      F    25    NaN   
6    Completed  Virginia              Arlington      2.0     NB    24     70   
7   Incomplete  Virginia                Ashburn      2.0      F    30    5.2   
8    Completed  Virginia             Woodbridge      2.0      M    24     71   
9   Incomplete     Texas                    NaN      4.0      M    26     71   
10   Completed  Maryland            Takoma Park      2.0      F    24     66   
11   Completed  Virginia                Fairfax      4.0      F    25     66   
12   Completed  Virginia              Chantilly      2.0      M    25     69   
13   Completed  Virginia                  Burke      2.0      F    23     66   
14   Completed  Maryland            Hyattsville      4.0      F    31     67   
15   Completed  Maryland              Baltimore      4.0      M    26     72   
16   Completed  Virginia              Arlington      2.0      M    40     68   
17   Completed  Virginia            Springfield      2.0      M   2 3     60   
18   Completed  Virginia             Woodbridge      4.0      M    25     67   
19   Completed  Virginia               Rustburg      4.0      F    22     67   
20   Completed  Virginia                Fairfax      4.0      M    23     74   
21   Completed  Virginia                 Reston      2.0      F    25     65   
22   Completed  Maryland              Baltimore      4.0      F    22     67   
23   Completed       NaN                    NaN      4.0      M    23     75   
24   Completed       NaN                    NaN      4.0      F    27     67   
25   Completed  Virginia                  Burke      2.0      M    28     68   
26   Completed  Virginia                Fairfax      2.0      M    23     72   
27   Completed  Virginia                  Burke      4.0      M    35     6l   
28   Completed  Maryland              Baltimore      4.0      F    27     64   
29   Completed  Virginia            Springfield      2.0      M    25     71   
30   Completed  Virginia                 Lorton      4.0      F    55     56   
31  Incomplete  Virginia                  Burke      2.0      F    23     70   
32   Completed  Virginia           Falls Church      2.0      M  1993     7O   
33   Completed       NaN                    NaN      4.0      M    23     73   
34   Completed  Maryland              Frederick      2.0      M    24     68   
35   Completed  Virginia                Fairfax      4.0      F    28     66   

                  Citizenship          Undergrad_Degree    GradYear  \
0                      Taiwan                 economics  12/31/2021   
1                       India                  Finance     5/1/2021   
2                       India     Computer Applications    6/1/2022   
3                 South Korea       Business Management  12/31/2021   
4                          US                       Yes    1/1/2021   
5                         NaN                       NaN         NaN   
6                    pakistan                   Finance   7/15/2022   
7                         NaN                       NaN         NaN   
8                       china                  Economic  12/15/2021   
9                 Uzbekistan    Economics with Finance          NaN   
10                        USA                       BIT    5/1/2022   
11                      India      Computer Engineering   5/31/2021   
12                    Romania                   Biology   8/26/2021   
13                      India                       C

## Handling Missing Values

### The 'Region' Column

In [109]:
df_clean['Region'].value_counts()

Region
Virginia    25
Maryland     6
Texas        1
Name: count, dtype: int64

In [110]:
df_clean['Region'].isnull().sum()

4

In [111]:
print('Total Number of entries:', len(df_clean))

Total Number of entries: 36


From this, it can be inferred that 'Virginia' is the most common region, therefore, we can use the mode to replace the missing values within this particular column.

In [112]:
mode_value = df_clean['Region'].mode()[0]
mode_value

'Virginia'

In [113]:
df_clean['Region'] = df_clean['Region'].fillna(mode_value)

Check for the missing values now

In [114]:
df_clean['Region'].isnull().sum()

0

In [115]:
df_clean['Region'].value_counts()

Region
Virginia    29
Maryland     6
Texas        1
Name: count, dtype: int64

### The 'District' Column

In [1]:
df_clean['City / Town / District'].value_counts()

NameError: name 'df_clean' is not defined

In [117]:
mode_district = df_clean['City / Town / District'].mode()[0]
mode_district

'Fairfax'

Eventhough this column does not have a value that's substantially larger than others, for simplicity, we can assume that the District is also the most commonly occurring one. But there is a problem with this approach. The City depends on the Region/State, and in some instances, the state is given and the city is not given, so if we use the mode value, there can be discrepancies in the data. Therefore, it would be better to drop this column entirely.

In [118]:
df_clean = df_clean.drop(columns=['City / Town / District'])

In [119]:
df_clean.head(10)

Status    Region  Section Gender Age Height  Citizenship  \
0   Completed  Virginia      4.0      F  23     63       Taiwan   
1   Completed  Virginia      2.0      F  24     59        India   
2   Completed  Virginia      2.0      M  40     70        India   
3   Completed  Virginia      2.0     No  26     66  South Korea   
4   Completed  Virginia      2.0      M  30     70           US   
5  Incomplete  Virginia      2.0      F  25    NaN          NaN   
6   Completed  Virginia      2.0     NB  24     70     pakistan   
7  Incomplete  Virginia      2.0      F  30    5.2          NaN   
8   Completed  Virginia      2.0      M  24     71        china   
9  Incomplete     Texas      4.0      M  26     71  Uzbekistan    

          Undergrad_Degree    GradYear   Laptop  Employement  Python Skills  \
0                economics  12/31/2021  MacBook  Not Working            2.0   
1                 Finance     5/1/2021  MacBook    Part Time            3.0   
2    Computer Applications    6/1/2022  MacBook    Full Time            2.0   
3      Business Management  12/31/2021  Windows  Not Working            3.0   
4                      Yes    1/1/2021  MacBook    Full Time            2.0   
5                      NaN         NaN      NaN          NaN            NaN   
6                  Finance   7/15/2022  Windows    Part Time            3.0   
7                      NaN         NaN      NaN          NaN            NaN   
8                 Economic  12/15/2021  Windows    Full Time            2.0   
9  Economics with Finance          NaN      NaN          NaN            NaN   

   R Skills  SQL Skills  GitHub Career in DS  \
0       1.0         2.0       1          Yes   
1       3.0         1.0       1           No   
2       1.0         3.0       0     Not sure   
3       4.0         2.0       0          Yes   
4       2.0         3.0       0          Yes   
5       NaN         NaN       1           No   
6       3.0         3.0       0           No   
7       NaN         NaN       1           No   
8       4.0         3.0       1           No   
9       NaN         NaN       0           No   

                                               Goals  
0                                                NaN  
1                                                NaN  
2  As part of my work, I have access to a lot of ...  
3                                                NaN  
4                                              Learn  
5                                                NaN  
6  To deduce meaningful interpretation for Financ...  
7                                                NaN  
8                                                NaN  
9                                                NaN

### The 'Age' Column

In [120]:
df_clean['Age'].head()

0    23
1    24
2    40
3    26
4    30
Name: Age, dtype: object

In [121]:
# Remove whitespaces from the 'Age' column
df_clean['Age'] = df_clean['Age'].str.replace(' ', '', regex=True)

# Convert the cleaned 'Age' column to integers
df_clean['Age'] = pd.to_numeric(df_clean['Age'], errors='coerce')

In [122]:
df_clean['Age'].head()

0    23
1    24
2    40
3    26
4    30
Name: Age, dtype: int64

In [123]:
# for the person who entered '2 3'
df_clean['Age'][18]

25

In [124]:
df_clean['Age'] = df_clean['Age'].apply(lambda x: x if x <= 90 else None)


In [125]:
df_clean['Age'].isnull().sum()

1

Filling with median because of the wide variety of ages in the dataset.

In [126]:
median_age = df_clean['Age'].median()
median_age

25.0

In [127]:
df_clean['Age'] = df_clean['Age'].fillna(median_age).astype(int)

In [128]:
df_clean['Age'].isnull().sum()

0

In [129]:
df_clean['Age'].head()

0    23
1    24
2    40
3    26
4    30
Name: Age, dtype: int32

### The 'Height' Column

In [130]:
df_clean['Height'].isnull().sum()

1

Since only one value is missing, and height is a number, therefore, the mean value can be chosen for this.

In [131]:
df_clean['Height'] = pd.to_numeric(df_clean['Height'], errors='coerce')

In [132]:
df_clean['Height']

0     63.0
1     59.0
2     70.0
3     66.0
4     70.0
5      NaN
6     70.0
7      5.2
8     71.0
9     71.0
10    66.0
11    66.0
12    69.0
13    66.0
14    67.0
15    72.0
16    68.0
17    60.0
18    67.0
19    67.0
20    74.0
21    65.0
22    67.0
23    75.0
24    67.0
25    68.0
26    72.0
27     NaN
28    64.0
29    71.0
30    56.0
31    70.0
32     NaN
33    73.0
34    68.0
35    66.0
Name: Height, dtype: float64

In [133]:
mean_height = df_clean.loc[df_clean['Height'] != 5.2, 'Height'].mean()
mean_height

67.625

In [134]:
df_clean['Height'] = df_clean['Height'].replace(5.2, mean_height)

In [135]:
df_clean['Height'] = df_clean['Height'].fillna(mean_height)
df_clean['Height'] = df_clean['Height'].astype(int)

In [136]:
df_clean.head(10)

Status    Region  Section Gender  Age  Height  Citizenship  \
0   Completed  Virginia      4.0      F   23      63       Taiwan   
1   Completed  Virginia      2.0      F   24      59        India   
2   Completed  Virginia      2.0      M   40      70        India   
3   Completed  Virginia      2.0     No   26      66  South Korea   
4   Completed  Virginia      2.0      M   30      70           US   
5  Incomplete  Virginia      2.0      F   25      67          NaN   
6   Completed  Virginia      2.0     NB   24      70     pakistan   
7  Incomplete  Virginia      2.0      F   30      67          NaN   
8   Completed  Virginia      2.0      M   24      71        china   
9  Incomplete     Texas      4.0      M   26      71  Uzbekistan    

          Undergrad_Degree    GradYear   Laptop  Employement  Python Skills  \
0                economics  12/31/2021  MacBook  Not Working            2.0   
1                 Finance     5/1/2021  MacBook    Part Time            3.0   
2    Computer Applications    6/1/2022  MacBook    Full Time            2.0   
3      Business Management  12/31/2021  Windows  Not Working            3.0   
4                      Yes    1/1/2021  MacBook    Full Time            2.0   
5                      NaN         NaN      NaN          NaN            NaN   
6                  Finance   7/15/2022  Windows    Part Time            3.0   
7                      NaN         NaN      NaN          NaN            NaN   
8                 Economic  12/15/2021  Windows    Full Time            2.0   
9  Economics with Finance          NaN      NaN          NaN            NaN   

   R Skills  SQL Skills  GitHub Career in DS  \
0       1.0         2.0       1          Yes   
1       3.0         1.0       1           No   
2       1.0         3.0       0     Not sure   
3       4.0         2.0       0          Yes   
4       2.0         3.0       0          Yes   
5       NaN         NaN       1           No   
6       3.0         3.0       0           No   
7       NaN         NaN       1           No   
8       4.0         3.0       1           No   
9       NaN         NaN       0           No   

                                               Goals  
0                                                NaN  
1                                                NaN  
2  As part of my work, I have access to a lot of ...  
3                                                NaN  
4                                              Learn  
5                                                NaN  
6  To deduce meaningful interpretation for Financ...  
7                                                NaN  
8                                                NaN  
9                                                NaN

In [137]:
df_clean['Height'].head(10)

0    63
1    59
2    70
3    66
4    70
5    67
6    70
7    67
8    71
9    71
Name: Height, dtype: int32

### The 'Citizenship' Column

In [138]:
df_clean['Citizenship'].value_counts()

Citizenship
India                        9
USA                          5
Taiwan                       2
United States                2
Uzbekistan                   2
India                        1
United States ; Pakistan     1
Nigeria                      1
Nigeria                      1
United States                1
Ethiopia                     1
China                        1
Romania                      1
Uzbekistan                   1
china                        1
pakistan                     1
US                           1
South Korea                  1
Vietnam                      1
Name: count, dtype: int64

There are multiple countries that are being repeated here. So we need to (1) remove the whitespaces, (2) create a mapping to resolve this.

In [139]:
# uniform cases
df_clean['Citizenship'] = df_clean['Citizenship'].str.title()

In [140]:
# remove whitespaces
df_clean['Citizenship'] = df_clean['Citizenship'].str.strip()


In [141]:
df_clean['Citizenship'].value_counts()

Citizenship
India                       10
Usa                          5
Uzbekistan                   3
United States                3
Taiwan                       2
China                        2
Nigeria                      2
South Korea                  1
Us                           1
Pakistan                     1
Romania                      1
Ethiopia                     1
United States ; Pakistan     1
Vietnam                      1
Name: count, dtype: int64

In [142]:
mapping = {
    'United States': 'USA',
    'Us': 'USA',
    'Usa': 'USA',
    'United States ; Pakistan': 'USA'
}

In [143]:
# replace with mappings
df_clean['Citizenship'] = df_clean['Citizenship'].replace(mapping)

In [144]:
df_clean['Citizenship'].value_counts()

Citizenship
India          10
USA            10
Uzbekistan      3
Taiwan          2
China           2
Nigeria         2
South Korea     1
Pakistan        1
Romania         1
Ethiopia        1
Vietnam         1
Name: count, dtype: int64

Now, check for the missing values.

In [145]:
df_clean['Citizenship'].isnull().sum()

2

In [146]:
mode_citizen = df_clean['Citizenship'].mode()
mode_citizen

0    India
1      USA
Name: Citizenship, dtype: object

Would be better to use the US as the citizenship since this data is from the US.

In [147]:
df_clean['Citizenship'] = df_clean['Citizenship'].fillna(mode_citizen[1])

In [148]:
df_clean['Citizenship'].isnull().sum()

0

In [149]:
df_clean['Citizenship'].value_counts()

Citizenship
USA            12
India          10
Uzbekistan      3
Taiwan          2
China           2
Nigeria         2
South Korea     1
Pakistan        1
Romania         1
Ethiopia        1
Vietnam         1
Name: count, dtype: int64

### The 'Undergrad Degree' Column

In [150]:
df_clean['Undergrad_Degree'].value_counts()

Undergrad_Degree
Economics                   3
Finance                     2
economics                   1
Statistics                  1
Criminology Law and Soci    1
Bachelor of engineering     1
BS accounting               1
B.Tech in CS                1
Bachelor of Engineering     1
Quantity Surveying          1
ECE                         1
Civil engineering           1
Systems Engineering         1
Mathematics & Statistics    1
Computer science            1
Computer Science            1
Engineering                 1
Accounting and ISOM         1
Finance                     1
Economics with Finance      1
Management                  1
Cse                         1
Biology                     1
Computer Engineering        1
BIT                         1
Economics with Finance      1
Economic                    1
Yes                         1
Business Management         1
Computer Applications       1
Psychology                  1
Name: count, dtype: int64

In [151]:
# uniform cases
df_clean['Undergrad_Degree'] = df_clean['Undergrad_Degree'].str.title()

In [152]:
# uniform cases
df_clean['Undergrad_Degree'] = df_clean['Undergrad_Degree'].str.strip()

In [153]:
df_clean['Undergrad_Degree'].value_counts()

Undergrad_Degree
Economics                   4
Finance                     3
Bachelor Of Engineering     2
Economics With Finance      2
Computer Science            2
Statistics                  1
Criminology Law And Soci    1
Bs Accounting               1
B.Tech In Cs                1
Quantity Surveying          1
Ece                         1
Civil Engineering           1
Systems Engineering         1
Mathematics & Statistics    1
Accounting And Isom         1
Engineering                 1
Management                  1
Cse                         1
Biology                     1
Computer Engineering        1
Bit                         1
Economic                    1
Yes                         1
Business Management         1
Computer Applications       1
Psychology                  1
Name: count, dtype: int64

In [154]:
def standardize_degree(degree):
    if isinstance(degree, str): 
        if 'Cs' in degree.title(): return 'Computer Science'
        elif 'Accounting' in degree.title(): return 'Accounting'
        elif 'Economic' in degree.title(): return 'Economics'
        elif 'Statistics' in degree.title(): return 'Mathematics'        
        elif ('Computer Engineering' in degree.title()) or ('Electrical Engineering' in degree.title()) or ('Ece' in degree.title()): return 'ECE' 
        elif 'Bachelor Of Engineering' == degree.title(): return 'Engineering'
        elif 'Management' in degree.title(): return 'Management'
    return degree  


df_clean['Undergrad_Degree'] = df_clean['Undergrad_Degree'].apply(standardize_degree)


In [155]:
df_clean['Undergrad_Degree'].value_counts()

Undergrad_Degree
Economics                   7
Computer Science            4
Finance                     3
Engineering                 3
Management                  2
ECE                         2
Accounting                  2
Mathematics                 2
Computer Applications       1
Yes                         1
Bit                         1
Biology                     1
Systems Engineering         1
Civil Engineering           1
Quantity Surveying          1
Criminology Law And Soci    1
Psychology                  1
Name: count, dtype: int64

In [156]:
df_clean['Undergrad_Degree'].head(10)

0                Economics
1                  Finance
2    Computer Applications
3               Management
4                      Yes
5                      NaN
6                  Finance
7                      NaN
8                Economics
9                Economics
Name: Undergrad_Degree, dtype: object

I notice someone answered 'Yes'. So to replace this we can use the mode.

In [157]:
mode_degree = df_clean['Undergrad_Degree'].mode()[0]
mode_degree

'Economics'

In [158]:
df_clean['Undergrad_Degree'] = df_clean['Undergrad_Degree'].replace('Yes', mode_degree)

In [159]:
df_clean['Undergrad_Degree'].isnull().sum()

2

Replace the null values with the mode as well.

In [160]:
df_clean['Undergrad_Degree'] = df_clean['Undergrad_Degree'].fillna(mode_degree)

In [161]:
df_clean['Undergrad_Degree'].isnull().sum()

0

In [162]:
df_clean['Undergrad_Degree'].head(10)

0                Economics
1                  Finance
2    Computer Applications
3               Management
4                Economics
5                Economics
6                  Finance
7                Economics
8                Economics
9                Economics
Name: Undergrad_Degree, dtype: object

### The 'Graduation Date' Column

In [163]:
df_clean['GradYear'].value_counts()

GradYear
12/31/2021    3
5/1/2022      3
12/20/2021    3
12/15/2021    2
5/1/2021      2
12/18/2021    2
1/1/2021      1
5/14/2021     1
5/15/2022     1
12/21/2021    1
12/30/2021    1
5/24/2023     1
12/11/2021    1
5/30/2021     1
12/19/2022    1
4/16/2022     1
7/15/2022     1
1/12/2021     1
5/31/2022     1
12/25/2021    1
8/26/2021     1
5/31/2021     1
6/1/2022      1
12/12/2022    1
Name: count, dtype: int64

In [164]:
df_clean['GradYear'].isnull().sum()

3

To make the data more readable, we can remove the dates, and only keep the month and year of graduation.

In [165]:
# convert to datetime
df_clean['GradYear'] = pd.to_datetime(df_clean['GradYear'], format='%m/%d/%Y')
# format to mm/yyyy
df_clean['GradYear'] = df_clean['GradYear'].dt.strftime('%m/%Y')

In [166]:
df_clean['GradYear'].value_counts()

GradYear
12/2021    14
05/2021     5
05/2022     5
01/2021     2
12/2022     2
06/2022     1
07/2022     1
08/2021     1
04/2022     1
05/2023     1
Name: count, dtype: int64

Now, this looks much more cleaner. We can use the mode now to fill the missing values. The reason for choosing mode is because there's a very large number of people graduating in 12/2021 so it makes sense for the missing three values to be on the same date as well.

In [167]:
mode_date = df_clean['GradYear'].mode()[0]
mode_date

'12/2021'

In [168]:
df_clean['GradYear'] = df_clean['GradYear'].fillna(mode_date)

In [169]:
df_clean['GradYear'].value_counts()

GradYear
12/2021    17
05/2021     5
05/2022     5
01/2021     2
12/2022     2
06/2022     1
07/2022     1
08/2021     1
04/2022     1
05/2023     1
Name: count, dtype: int64

In [170]:
df_clean['GradYear'].isnull().sum()

0

### The 'Laptop' Column

In [171]:
df_clean['Laptop'].value_counts()

Laptop
Windows           20
MacBook            8
WindowsMacBook     3
Other              1
Name: count, dtype: int64

This shows that there are 4 missing values, and one incorrect value 'WindowsMacBook' that should be replaced with 'Both'.

In [172]:
df_clean['Laptop'] = df_clean['Laptop'].str.strip()

In [173]:
# replace WindowsMacBook with Both
df_clean['Laptop'] = df_clean['Laptop'].replace('WindowsMacBook', 'Both')

In [174]:
# replace '' with NaN
df_clean['Laptop'] = df_clean['Laptop'].replace('', None)

In [175]:
df_clean['Laptop'].value_counts()

Laptop
Windows    20
MacBook     8
Both        3
Other       1
Name: count, dtype: int64

Now, we can replace with NaN with mode because here as well 'Windows' is the clear majority so mode is a safe choice.

In [176]:
mode_laptop = df_clean['Laptop'].mode()[0]
mode_laptop

'Windows'

In [177]:
df_clean['Laptop'] = df_clean['Laptop'].fillna(mode_laptop)

In [178]:
df_clean['Laptop'].isnull().sum()

0

In [179]:
df_clean['Laptop'].value_counts()

Laptop
Windows    24
MacBook     8
Both        3
Other       1
Name: count, dtype: int64

### The 'Employement' Column

In [180]:
df_clean['Employement'].value_counts()

Employement
Not Working    17
Full Time      11
Part Time       5
Name: count, dtype: int64

Replace the '' with None to completely identify the missing values.

In [181]:
df_clean['Employement'] = df_clean['Employement'].replace('', None)

In [182]:
df_clean['Employement'].value_counts()

Employement
Not Working    17
Full Time      11
Part Time       5
Name: count, dtype: int64

Now use the mode to fill in the missing values.

In [183]:
median_job = df_clean['Employement'].mode()[0]
median_job

'Not Working'

In [184]:
df_clean['Employement'] = df_clean['Employement'].fillna(median_job)

In [185]:
df_clean['Employement'].isnull().sum()

0

In [186]:
df_clean['Employement'].value_counts()

Employement
Not Working    20
Full Time      11
Part Time       5
Name: count, dtype: int64

### The 'Python Skills' Column

In [187]:
df_clean['Python Skills'].value_counts()

Python Skills
3.0    15
2.0    12
1.0     4
4.0     2
Name: count, dtype: int64

Replace '' with None, so it's clear that it's a missing value.

In [188]:
df_clean['Python Skills'] = df_clean['Python Skills'].replace('', None)

Now, to fill in the missing values, we can use the median as this is ordinal data with discrete values.

In [189]:
median_skill = df_clean['Python Skills'].median()
median_skill

3.0

In [190]:
df_clean['Python Skills'] = df_clean['Python Skills'].fillna(median_skill).astype(int)

In [191]:
df_clean['Python Skills'].value_counts()

Python Skills
3    18
2    12
1     4
4     2
Name: count, dtype: int64

### The 'R Skills' Column

Will follow the same strategy as I did for the Python Skills' column here as well as the data is same.

In [192]:
df_clean['R Skills'].value_counts()

R Skills
3.0    9
1.0    8
2.0    8
4.0    7
5.0    1
Name: count, dtype: int64

In [193]:
# get rid of '' values
df_clean['R Skills'] = df_clean['R Skills'].replace('', None)

In [194]:
# fill missing values
median_R = df_clean['R Skills'].median()
median_R

3.0

In [195]:
df_clean['R Skills'] = df_clean['R Skills'].fillna(median_R).astype(int)

In [196]:
df_clean['R Skills'].value_counts()

R Skills
3    12
1     8
2     8
4     7
5     1
Name: count, dtype: int64

### The 'SQL Skills' Column

Following the same procedure as Python and R skills columns.

In [197]:
df_clean['SQL Skills'].value_counts()

SQL Skills
3.0    15
2.0    10
1.0     5
4.0     3
Name: count, dtype: int64

In [198]:
# get rid of '' values
df_clean['SQL Skills'] = df_clean['SQL Skills'].replace('', None)

In [199]:
# fill missing values
median_SQL = df_clean['SQL Skills'].median()
median_SQL


3.0

In [200]:
df_clean['SQL Skills'] = df_clean['SQL Skills'].fillna(median_SQL).astype(int)

In [201]:
df_clean['SQL Skills'].value_counts()

SQL Skills
3    18
2    10
1     5
4     3
Name: count, dtype: int64

## Final Cleaned Data

All missing values have now been dealt with, except Q15 which contains open-ended text.

In [69]:
df_clean.head()

NameError: name 'df_clean' is not defined

Now, we don't need the completed column, so I will drop it.

In [203]:
df_clean = df_clean.drop(columns='Status')

In [204]:
df_clean.isnull().sum()

Region              0
Section             0
Gender              0
Age                 0
Height              0
Citizenship         0
Undergrad_Degree    0
GradYear            0
Laptop              0
Employement         0
Python Skills       0
R Skills            0
SQL Skills          0
GitHub              0
Career in DS        0
Goals               9
dtype: int64

In [205]:
df_clean

Region  Section Gender  Age  Height  Citizenship  \
0   Virginia      4.0      F   23      63       Taiwan   
1   Virginia      2.0      F   24      59        India   
2   Virginia      2.0      M   40      70        India   
3   Virginia      2.0     No   26      66  South Korea   
4   Virginia      2.0      M   30      70          USA   
5   Virginia      2.0      F   25      67          USA   
6   Virginia      2.0     NB   24      70     Pakistan   
7   Virginia      2.0      F   30      67          USA   
8   Virginia      2.0      M   24      71        China   
9      Texas      4.0      M   26      71   Uzbekistan   
10  Maryland      2.0      F   24      66          USA   
11  Virginia      4.0      F   25      66        India   
12  Virginia      2.0      M   25      69      Romania   
13  Virginia      2.0      F   23      66        India   
14  Maryland      4.0      F   31      67        China   
15  Maryland      4.0      M   26      72   Uzbekistan   
16  Virginia      2.0      M   40      68          USA   
17  Virginia      2.0      M   23      60        India   
18  Virginia      4.0      M   25      67     Ethiopia   
19  Virginia      4.0      F   22      67          USA   
20  Virginia      4.0      M   23      74          USA   
21  Virginia      2.0      F   25      65        India   
22  Maryland      4.0      F   22      67          USA   
23  Virginia      4.0      M   23      75      Nigeria   
24  Virginia      4.0      F   27      67          USA   
25  Virginia      2.0      M   28      68       Taiwan   
26  Virginia      2.0      M   23      72        India   
27  Virginia      4.0      M   35      67      Nigeria   
28  Maryland      4.0      F   27      64        India   
29  Virginia      2.0      M   25      71        India   
30  Virginia      4.0      F   55      56          USA   
31  Virginia      2.0      F   23      70        India   
32  Virginia      2.0      M   25      67          USA   
33  Virginia      4.0      M   23      73          USA   
34  Maryland      2.0      M   24      68      Vietnam   
35  Virginia      4.0      F   28      66   Uzbekistan   

            Undergrad_Degree GradYear   Laptop  Employement  Python Skills  \
0                  Economics  12/2021  MacBook  Not Working              2   
1                    Finance  05/2021  MacBook    Part Time              3   
2      Computer Applications  06/2022  MacBook    Full Time              2   
3                 Management  12/2021  Windows  Not Working              3   
4                  Economics  01/2021  MacBook    Full Time              2   
5                  Economics  12/2021  Windows  Not Working              3   
6                    Finance  07/2022  Windows    Part Time              3   
7                  Economics  12/2021  Windows  Not Working              3   
8                  Economics  12/2021  Windows    Full Time              2   
9                  Economics  12/2021  Windows  Not Working              3   
10                       Bit  05/2022  Windows  Not Working              1   
11                       ECE  05/2021  Windows  Not Working              4   
12                   Biology  08/2021  MacBook    Part Time              2   
13          Computer Science  12/2021  Windows  Not Working              3   
14                Management  12/2021  Windows  Not Working              3   
15                 Economics  05/2022  Windows  Not Working              2   
16                 Economics  05/2022     Both    Full Time              3   
17               Engineering  01/2021  Windows  Not Working              4   
18                Accounting  12/2021  Windows  Not Working              1   
19               Mathematics  12/2021  MacBook    Part Time              2   
20          Computer Science  05/2022  Windows    Full Time              2   
21          Computer Science  04/2022  Windows  Not Working              2   
22               Mathematics  12/2021     Both    Part Time      

In [206]:
# saving the CSV file
df_clean.to_csv('cleaned_hw1_data.csv', index=False, header=True)

# Part 3: Statistical Summaries

In [207]:
# read the cleaned data file
df = pd.read_csv('cleaned_hw1_data.csv')

In [208]:
df.head(10)

Region  Section Gender  Age  Height  Citizenship       Undergrad_Degree  \
0  Virginia      4.0      F   23      63       Taiwan              Economics   
1  Virginia      2.0      F   24      59        India                Finance   
2  Virginia      2.0      M   40      70        India  Computer Applications   
3  Virginia      2.0     No   26      66  South Korea             Management   
4  Virginia      2.0      M   30      70          USA              Economics   
5  Virginia      2.0      F   25      67          USA              Economics   
6  Virginia      2.0     NB   24      70     Pakistan                Finance   
7  Virginia      2.0      F   30      67          USA              Economics   
8  Virginia      2.0      M   24      71        China              Economics   
9     Texas      4.0      M   26      71   Uzbekistan              Economics   

  GradYear   Laptop  Employement  Python Skills  R Skills  SQL Skills  GitHub  \
0  12/2021  MacBook  Not Working              2         1           2       1   
1  05/2021  MacBook    Part Time              3         3           1       1   
2  06/2022  MacBook    Full Time              2         1           3       0   
3  12/2021  Windows  Not Working              3         4           2       0   
4  01/2021  MacBook    Full Time              2         2           3       0   
5  12/2021  Windows  Not Working              3         3           3       1   
6  07/2022  Windows    Part Time              3         3           3       0   
7  12/2021  Windows  Not Working              3         3           3       1   
8  12/2021  Windows    Full Time              2         4           3       1   
9  12/2021  Windows  Not Working              3         3           3       0   

  Career in DS                                              Goals  
0          Yes                                                NaN  
1           No                                                NaN  
2     Not sure  As part of my work, I have access to a lot of ...  
3          Yes                                                NaN  
4          Yes                                              Learn  
5           No                                                NaN  
6           No  To deduce meaningful interpretation for Financ...  
7           No                                                NaN  
8           No                                                NaN  
9           No                                                NaN

Drop the Q15 because it has text.

In [209]:
df = df.drop(columns='Goals')

Let's see what's inside this cleaned data.

In [210]:
df.describe(include='all')

Region    Section Gender        Age     Height Citizenship  \
count         36  36.000000     36  36.000000  36.000000          36   
unique         3        NaN      4        NaN        NaN          11   
top     Virginia        NaN      M        NaN        NaN         USA   
freq          29        NaN     18        NaN        NaN          12   
mean         NaN   2.888889    NaN  27.000000  67.555556         NaN   
std          NaN   1.007905    NaN   6.493953   3.952897         NaN   
min          NaN   2.000000    NaN  22.000000  56.000000         NaN   
25%          NaN   2.000000    NaN  23.000000  66.000000         NaN   
50%          NaN   2.000000    NaN  25.000000  67.000000         NaN   
75%          NaN   4.000000    NaN  27.250000  70.000000         NaN   
max          NaN   4.000000    NaN  55.000000  75.000000         NaN   

       Undergrad_Degree GradYear   Laptop  Employement  Python Skills  \
count                36       36       36           36      36.000000   
unique               16       10        4            3            NaN   
top           Economics  12/2021  Windows  Not Working            NaN   
freq                 10       17       24           20            NaN   
mean                NaN      NaN      NaN          NaN       2.500000   
std                 NaN      NaN      NaN          NaN       0.774597   
min                 NaN      NaN      NaN          NaN       1.000000   
25%                 NaN      NaN      NaN          NaN       2.000000   
50%                 NaN      NaN      NaN          NaN       3.000000   
75%                 NaN      NaN      NaN          NaN       3.000000   
max                 NaN      NaN      NaN          NaN       4.000000   

         R Skills  SQL Skills     GitHub Career in DS  
count   36.000000   36.000000  36.000000           36  
unique        NaN         NaN        NaN            3  
top           NaN         NaN        NaN          Yes  
freq          NaN         NaN        NaN           17  
mean     2.583333    2.527778   0.333333          NaN  
std      1.130739    0.844685   0.478091          NaN  
min      1.000000    1.000000   0.000000          NaN  
25%      2.000000    2.000000   0.000000          NaN  
50%      3.000000    3.000000   0.000000          NaN  
75%      3.000000    3.000000   1.000000          NaN  
max      5.000000    4.000000   1.000000          NaN

Just from this table it can be inferred that most of the people who took this class were Economics majors who wanted to pursue a career in data analytics/data sciene.

In [211]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36 entries, 0 to 35
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Region            36 non-null     object 
 1   Section           36 non-null     float64
 2   Gender            36 non-null     object 
 3   Age               36 non-null     int64  
 4   Height            36 non-null     int64  
 5   Citizenship       36 non-null     object 
 6   Undergrad_Degree  36 non-null     object 
 7   GradYear          36 non-null     object 
 8   Laptop            36 non-null     object 
 9   Employement       36 non-null     object 
 10  Python Skills     36 non-null     int64  
 11  R Skills          36 non-null     int64  
 12  SQL Skills        36 non-null     int64  
 13  GitHub            36 non-null     int64  
 14  Career in DS      36 non-null     object 
dtypes: float64(1), int64(6), object(8)
memory usage: 4.3+ KB


## For Individual Questions

### For Q1 - Section

In [212]:
df['Section'].value_counts()

Section
2.0    20
4.0    16
Name: count, dtype: int64

Both sections were almost equally represented, with Section 004 receiving a slightly higher number of responses.

### For Q2 - Gender

In [213]:
df['Gender'].value_counts()

Gender
M     18
F     16
No     1
NB     1
Name: count, dtype: int64

This represents that there was almost an equal gender representation among the students who took this class - assuming that everyone who was enrolled in the course filled out the survey too.

### For Q3 - Age

In [214]:
df['Age'].value_counts()

Age
23    8
25    7
24    5
26    3
40    2
30    2
22    2
27    2
28    2
31    1
35    1
55    1
Name: count, dtype: int64

In [215]:
df['Age'].describe()

count    36.000000
mean     27.000000
std       6.493953
min      22.000000
25%      23.000000
50%      25.000000
75%      27.250000
max      55.000000
Name: Age, dtype: float64

From the course code, and the fact that there was a question about the undergraduate degree, it seems like this is graduate level course. The ages of the students enrolled in it confirms this, with minimum 22 years old and then a wide rane of ages - upto 55 - it is evident that this is infact a graduate class. Because if it was an undergraduate class, the minimum age would be around 19-20 and their wouldn't have been this large of a standard deviation (6.5 years) and mean (27 years) as it is currently.

### For Q4 - Height

In [216]:
df['Height'].value_counts()

Height
67    9
66    5
70    4
71    3
68    3
72    2
63    1
59    1
69    1
60    1
74    1
65    1
75    1
64    1
56    1
73    1
Name: count, dtype: int64

In [217]:
df['Height'].describe()

count    36.000000
mean     67.555556
std       3.952897
min      56.000000
25%      66.000000
50%      67.000000
75%      70.000000
max      75.000000
Name: Height, dtype: float64

The average height is approximately 67.56 inches, indicating a typical range for adult height, likely centered around normal distribution values. 75% of the individuals are shorter than 70 inches, which shows that the majority of the sample is clustered within a typical height range.

### For Q5 - Country of Citizenship

In [218]:
df['Citizenship'].value_counts()

Citizenship
USA            12
India          10
Uzbekistan      3
Taiwan          2
China           2
Nigeria         2
South Korea     1
Pakistan        1
Romania         1
Ethiopia        1
Vietnam         1
Name: count, dtype: int64

This shows that the class was pretty diverse, with 2/3 of the people from outside the US, and from them, there was a vast majority of Asian people, especially due the substantial number of Indian students.
It means that the course was appealing to a lot of people, regardless of their background, probably because of the possible job prospects in a lucrative field. This also explains the number of Asian people in the course, because people normally immigrate from this region in search of high paying jobs, and studying courses like this helps them pave the path.

### For Q6 - Undergraduate Degree

In [219]:
df['Undergrad_Degree'].value_counts()

Undergrad_Degree
Economics                   10
Computer Science             4
Finance                      3
Engineering                  3
Management                   2
ECE                          2
Accounting                   2
Mathematics                  2
Computer Applications        1
Bit                          1
Biology                      1
Systems Engineering          1
Civil Engineering            1
Quantity Surveying           1
Criminology Law And Soci     1
Psychology                   1
Name: count, dtype: int64

Out of all the students, almost half of them have a STEM background, particularly from CS and ECE related majors which makes sense as data analysis is extremely important for engineering and science. The other half is from an economics + finance background which also makes sense because data science is needed in their field as well to take data-driven decisions. 

Interestingly, there are a few students who are not from and Econ or STEM background. This just shows how robust and necessary the study of Data Science is because people from diverse backgrounds study it and use it in their work.

### For Q7 - Graduation Date 

In [220]:
df['GradYear'].value_counts()

GradYear
12/2021    17
05/2021     5
05/2022     5
01/2021     2
12/2022     2
06/2022     1
07/2022     1
08/2021     1
04/2022     1
05/2023     1
Name: count, dtype: int64

In [221]:
df['GradYear'].describe()

count          36
unique         10
top       12/2021
freq           17
Name: GradYear, dtype: object

It shows that most people graduate in Dec 2021, which means that most of them are from the same batch/level. However, most students graduate after 2021 which means that the class has people from different batches, and the program also caters people who are enrolled in the MS program at different paces - full-time or part-time which makes sense because there is a good number of people who is also working part time or full time according to one of the columns.

### For Q8 - Laptop

In [224]:
df['Laptop'].value_counts()

Laptop
Windows    24
MacBook     8
Both        3
Other       1
Name: count, dtype: int64

In [223]:
df['Laptop'].describe()

count          36
unique          4
top       Windows
freq           24
Name: Laptop, dtype: object

Most students have a windows laptop (24 + 3) possibly because Windows offers a wide range of data science tools and is generally easy to set up for data science tasks, making it a practical choice for students who need reliable and widely-supported software. And the other reason is that MacBook is more expensive than a Windows laptop, and might not be affordable for window. The other would certainly be a Linux-based OS which most people can not use because of its heavy dependence of the command line termina.

### For Q9 - Employement

In [225]:
df['Employement'].value_counts()

Employement
Not Working    20
Full Time      11
Part Time       5
Name: count, dtype: int64

In [226]:
df['Employement'].describe()

count              36
unique              3
top       Not Working
freq               20
Name: Employement, dtype: object

This shows that most people are not working, and are full time students, while there are 16 people who have some sort of a job while studying. This adds up if you look at the Graduation column where those not working graduate early in Dec 2021.

In [238]:
filtered_df = df[(df['Employement'] == 'Not Working') & (df['GradYear'].str.contains('2021'))]
filtered_df.shape[0]

15

From this we can see that there are 15 out of 20 people who are not working at all graduate in 2021, which means if you study full you can graduate earlier because you'll be able to give more time to your classes.
But this also shows that the university supports people who wish to continue their job while studying as there's a substantial number of people who are employed - and most of them full time.

### For Q10 - Python Skills

In [239]:
df['Python Skills'].value_counts()

Python Skills
3    18
2    12
1     4
4     2
Name: count, dtype: int64

In [240]:
df['Python Skills'].describe()

count    36.000000
mean      2.500000
std       0.774597
min       1.000000
25%       2.000000
50%       3.000000
75%       3.000000
max       4.000000
Name: Python Skills, dtype: float64

The majority of students rate their Python skills as average (3) or some familiarity (2), with fewer students identifying as either complete beginners (1) or frequent users (4), and no one considering themselves fluent (5). This distribution can guide the course content focus towards solidifying intermediate skills while offering support for both lower and upper ends of the skill spectrum.

### For Q11 - R Skills

In [242]:
df['R Skills'].value_counts()

R Skills
3    12
1     8
2     8
4     7
5     1
Name: count, dtype: int64

In [241]:
df['R Skills'].describe()

count    36.000000
mean      2.583333
std       1.130739
min       1.000000
25%       2.000000
50%       3.000000
75%       3.000000
max       5.000000
Name: R Skills, dtype: float64

The R skills among students vary significantly, with a concentration around the intermediate level. The data indicates that while most students have a working knowledge of R, there is a need for support at both ends of the spectrum—providing more foundational resources for beginners and advanced challenges for the few who are further along. Understanding these varying skill levels can help tailor the course content to better meet the needs of the students.
In comparison to Python though, students feel more skilled in R.

### For Q12 - SQL Skills

In [243]:
df['SQL Skills'].value_counts()

SQL Skills
3    18
2    10
1     5
4     3
Name: count, dtype: int64

In [244]:
df['SQL Skills'].describe()

count    36.000000
mean      2.527778
std       0.844685
min       1.000000
25%       2.000000
50%       3.000000
75%       3.000000
max       4.000000
Name: SQL Skills, dtype: float64

The data indicates that most students have intermediate SQL skills, with relatively few at the beginner or advanced ends of the spectrum. The moderate variability suggests a need for a balanced approach to instruction, focusing on solidifying foundational skills for beginners while providing more complex SQL challenges to help intermediate students advance. This distribution can guide how course materials and learning support are tailored to better address the varying skill levels present in the class.

It also appears that students feel the most confident with SQL, followed by R and then Python. Which could also be because of the difficulty level of each language.

### For Q13 - GitHub

In [245]:
df['GitHub'].value_counts()

GitHub
0    24
1    12
Name: count, dtype: int64

In [246]:
df['GitHub'].describe()

count    36.000000
mean      0.333333
std       0.478091
min       0.000000
25%       0.000000
50%       0.000000
75%       1.000000
max       1.000000
Name: GitHub, dtype: float64

The data indicates a split in GitHub usage among students, with a majority not using the platform - which shouldn't be the case in an MS level CS class. Addressing this disparity by providing GitHub training or integrating version control practices into the curriculum could enhance students' ability to manage and collaborate on projects effectively.

### For Q14 - Career in Data Science

In [247]:
df['Career in DS'].value_counts()

Career in DS
Yes         17
Not sure    10
No           9
Name: count, dtype: int64

In [248]:
df['Career in DS'].describe()

count      36
unique      3
top       Yes
freq       17
Name: Career in DS, dtype: object

The data reveals a predominantly positive outlook toward careers in data science among students, with a substantial number also being uncertain about their future in the field. Addressing the varying levels of interest and uncertainty through targeted educational and career support initiatives could enhance students' career readiness and satisfaction with their career choices.